core/src/Streamly/Data/Array/Unboxed.hs

```diff
@@ -0,0 +1,105 @@
+{-# OPTIONS_GHC -Wno-deprecations #-}
+#include "inline.hs"
+
+-- |
+-- Module      : Streamly.Data.Array.Foreign
+-- Copyright   : (c) 2019 Composewell Technologies
+--
+-- License     : BSD3
+-- Maintainer  : streamly@composewell.com
+-- Stability   : released
+-- Portability : GHC
+--
+-- This module provides immutable arrays in pinned memory (non GC memory)
+-- suitable for long lived data storage, random access and for interfacing with
+-- the operating system.
+--
+-- Arrays in this module are chunks of pinned memory that hold a sequence of
+-- 'Storable' values of a given type, they cannot store non-serializable data
+-- like functions.  Once created an array cannot be modified.  Pinned memory
+-- allows efficient buffering of long lived data without adding any impact to
+-- GC. One array is just one pointer visible to GC and it does not have to be
+-- copied across generations.  Moreover, pinned memory allows communication
+-- with foreign consumers and producers (e.g. file or network IO) without
+-- copying the data.
```

Comment:
For the time being, add an XXX comment stating to change the documentation and open an issue about it.